lib/Dialect/FIRRTL/Transforms/Dedup.cpp

```diff
@@ -71,20 +82,28 @@ struct StructuralHasher {
     nonessentialAttributes.insert(StringAttr::get(context, "inner_sym"));
   };
 
-  std::array<uint8_t, 32> hash(FModuleLike module) {
+  // This is a cached "portTypes" string attr.
+  StringAttr portTypesAttr;
+
+  // This is a cached "moduleName" string attr.
+  StringAttr moduleNameAttr;
+
+  // This is a set of every attribute we should ignore.
+  DenseSet<Attribute> nonessentialAttributes;
+};
+
+struct StructuralHasher {
+  explicit StructuralHasher(const StructuralHasherSharedConstants &constants)
+      : constants(constants){};
+
+  Hash hash(FModuleLike module) {
     update(&(*module));
     auto hash = sha.final();
-    reset();
-    return hash;
+    auto array = mlir::ArrayAttr::get(module.getContext(), referredModuleNames);
+    return {hash, array};
   }
 
 private:
-  void reset() {
-    currentIndex = 0;
-    indexes.clear();
-    sha.init();
-  }
-
```

Comment:
I removed `reset` since `Hasher` is initialized every time. It's a bit inefficient and ideally we should be able to create `Hasher` per thread, but I think currently MLIR threading library doesn't provide such API.